File: nodejs/tests/pressure/bmp581_test.js
'use strict';
const { I2CConnection } = require('../../packages/periph/src/connection/i2c');
const { BMP581Minimal, BMP581Full } = require('../../packages/periph/src/chips/pressure/bmp581');

const I2C_BUS  = parseInt(process.env.I2C_BUS  || '1',  10);
const I2C_ADDR = parseInt(process.env.I2C_ADDR || '0x46', 16);

let passed = 0, failed = 0;

async function main() {
    const connection = new I2CConnection(I2C_BUS, I2C_ADDR);
    const bmp = new BMP581Minimal(connection);

    // Software decode smoke: 0x1000 -> 0.0625 °C, 0x04 -> 0.0625 Pa.
    let rawT = 0x1000;
    let rawP = 0x04;
    if (rawT & 0x800000) rawT -= 0x1000000;
    if (rawP & 0x800000) rawP -= 0x1000000;
    const t = rawT / 65536.0;
    const p = rawP / 64.0;
    if (Math.abs(t - 0.0625) < 1e-6) { console.log('PASS temperature_decode'); passed++; }
    else { console.log('FAIL temperature_decode: got ' + t); failed++; }
    if (Math.abs(p - 0.0625) < 1e-6) { console.log('PASS pressure_decode'); passed++; }
    else { console.log('FAIL pressure_decode: got ' + p); failed++; }

    const bmpFull = new BMP581Full(connection);
    await bmpFull.setMode(BMP581Full.MODE_NORMAL);
    if (bmpFull._pwrMode === 1) { console.log('PASS set_mode'); passed++; }
    else { console.log('FAIL set_mode'); failed++; }

    await bmpFull.configure(0x17, 4, 2, true);
    if (bmpFull._odr === 0x17 && bmpFull._osrP === 4 && bmpFull._osrT === 2) {
        console.log('PASS configure'); passed++;
    } else { console.log('FAIL configure'); failed++; }

    await bmpFull.setIirFilter(BMP581Full.IIR_COEFF_3, BMP581Full.IIR_BYPASS);
    console.log('PASS set_iir_filter'); passed++;
    await bmpFull.enableDrdyInterrupt(true);
    console.log('PASS enable_drdy_interrupt'); passed++;
    await bmpFull.configureFifo(BMP581Full.FIFO_BOTH, BMP581Full.FIFO_STREAM, 8);
    console.log('PASS configure_fifo'); passed++;
    await bmpFull.setOorThreshold(110000, 200, 1);
    console.log('PASS set_oor_threshold'); passed++;

    const alt = await bmpFull.altitude();
    if (alt >= -500 && alt <= 9000) { console.log('PASS altitude'); passed++; }
    else { console.log('FAIL altitude'); failed++; }

    await connection.close();
    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
}

main();

File: python/tests/pressure/bmp581_test.py
import _testconfig as cfg
from periph.connection.i2c_micropython import I2CConnection
from periph.chips.pressure.bmp581 import BMP581Minimal, BMP581Full
from machine import Pin

passed = 0
failed = 0

i2c = I2C(cfg.I2C_ID, sda=Pin(cfg.SDA), scl=Pin(cfg.SCL), freq=cfg.FREQ)
connection = I2CConnection(i2c, cfg.ADDR)

bmp = BMP581Minimal(connection)

raw_t = 0x007E00
raw_p = 0x003D65
if raw_t & 0x800000:
    raw_t -= 0x1000000
if raw_p & 0x800000:
    raw_p -= 0x1000000
t = raw_t / 65536.0
p = raw_p / 64.0
if abs(t - 0.0625) < 1e-6:
    print('PASS temperature_decode')
    passed += 1
else:
    print('FAIL temperature_decode: got {}'.format(t))
    failed += 1
if abs(p - 0.0625) < 1e-6:
    print('PASS pressure_decode')
    passed += 1
else:
    print('FAIL pressure_decode: got {}'.format(p))
    failed += 1

bmp_full = BMP581Full(connection)
bmp_full.set_mode(BMP581Full.MODE_NORMAL)
if bmp_full._pwr_mode == 1:
    print('PASS set_mode')
    passed += 1
else:
    print('FAIL set_mode: got {}'.format(bmp_full._pwr_mode))
    failed += 1

bmp_full.configure(odr=0x17, osr_p=4, osr_t=2, press_en=True)
if bmp_full._odr == 0x17 and bmp_full._osr_p == 4 and bmp_full._osr_t == 2:
    print('PASS configure')
    passed += 1
else:
    print('FAIL configure: got odr={} osr_p={} osr_t={}'.format(bmp_full._odr, bmp_full._osr_p, bmp_full._osr_t))
    failed += 1

bmp_full.set_iir_filter(BMP581Full.IIR_COEFF_3, BMP581Full.IIR_BYPASS)
print('PASS set_iir_filter (no exception)')
passed += 1

bmp_full.enable_drdy_interrupt(True)
print('PASS enable_drdy_interrupt (no exception)')
passed += 1

bmp_full.configure_fifo(BMP581Full.FIFO_BOTH, mode=BMP581Full.FIFO_STREAM, threshold=8)
print('PASS configure_fifo (no exception)')
passed += 1

bmp_full.set_oor_threshold(110000, 200, 1)
print('PASS set_oor_threshold (no exception)')
passed += 1

alt = bmp_full.altitude()
if alt >= -500 and alt <= 9000:
    print('PASS altitude')
    passed += 1
else:
    print('FAIL altitude: got {}'.format(alt))
    failed += 1

print('===DONE: {} passed, {} failed==='.format(passed, failed))